consensus/polybft/fsm.go

```diff
@@ -242,13 +237,47 @@ func (f *fsm) Validate(proposal []byte) error {
 		)
 	}
 
-	blockExtra, err := GetIbftExtra(block.Header.ExtraData)
+	currentExtra, err := GetIbftExtra(block.Header.ExtraData)
 	if err != nil {
 		return fmt.Errorf("cannot get extra data:%w", err)
 	}
 
+	parentExtra, err := GetIbftExtra(f.parent.ExtraData)
+	if err != nil {
+		return err
+	}
+
+	currentValidators := f.validators.Accounts()
+	nextValidators := f.validators.Accounts()
+
+	if f.isEndOfEpoch {
+		if err := f.blockBuilder.Reset(); err != nil {
+			return fmt.Errorf("failed to initialize block builder: %w", err)
+		}
+
+		// we need to write commit epoch transaction to intermediate state,
+		// in order to be able to query next validator set
+		if err := f.blockBuilder.WriteTx(block.Transactions[0]); err != nil {
+			return err
+		}
+
+		nextValidators, err = f.getCurrentValidators(f.blockBuilder.GetState())
```

Comment:
Since `ProcessBlock` that is called at the end of the `fsm.Validate` function already executes the transaction, it would be nice to use that to validate this as well. Maybe expand `ProcessBlock` to return a new object:
```
ProcessedBlock {
  FulLBlock *types.FullBlock
  Transition *Transition 
}
```
That way you can use it to have this validation as well, without the need to double apply the same transaction.

Reply:
I've chosen to take different approach: instead of providing `Transition` object, I've isolated scope of transition instance through callback function. Otherwise we would enable client code to alter the state, which could be dangerous.

Fixed https://github.com/0xPolygon/polygon-edge/pull/1160/commits/fbe0c20c52382b2b0b6528d5dd68f5b25ec7a971.